### PR TITLE
Fix rustc sysroot in systems using CAS

### PR DESCRIPTION
**Pull request by rcvalle**
Change filesearch::get_or_default_sysroot() to check if sysroot is found using env::args().next() if rustc in argv[0] is a symlink; otherwise, or if it is not found, use env::current_exe() to imply sysroot. This makes the rustc binary able to locate Rust libraries in systems using content-addressable storage (CAS).

### CODE REVIEW

**Comment by jyn514** [comment]
@rcvalle what is content-addressable storage? Do you mean like a `nix` store or something?

**Comment by Mark-Simulacrum** [comment]
I don't think I can effectively review this patch without more information -- switching away from current_exe feels like the wrong approach though. If it was a "true" content-addressable storage system I would expect paths to be largely meaningless, so the popping that is still present wouldn't work really either...?

**Comment by rcvalle** [comment]
> @rcvalle what is content-addressable storage? Do you mean like a `nix` store or something?

Content-addressable storage (CAS) is a method of storing information so it can be retrieved based on its content (usually using a hashing function appropriate for CAS) instead of its location. (Git itself is an userspace CAS filesystem.)

**Comment by rcvalle** [comment]
> I don't think I can effectively review this patch without more information -- switching away from current_exe feels like the wrong approach though. If it was a "true" content-addressable storage system I would expect paths to be largely meaningless, so the popping that is still present wouldn't work really either...?

Sorry, I should have provided more information. We use CAS in our build systems with symlink forests to replicate the locations files originally belong. However, since both env::current_exe() and fs::canonicalize() follow symlinks/canonicalize the components of the path, the symlinks were being followed to the CAS locations, making the rustc binary unable to locate its distributed libraries.

**Comment by Mark-Simulacrum** [comment]
Hm, ok. I'm not sure I'm willing to land this change -- it seems like there are cases where that canonicalization would be necessary. Generally the sysroot discovery is a bit ad-hoc in rustc and not super well polished. I think my suggestion would be that we try both - it looks like perhaps the [sysroot_candidates](https://github.com/rust-lang/rust/blob/068320b39e3e4839d832b3aa71fa910ba170673b/compiler/rustc_interface/src/util.rs#L276) function can be extended to also visit the non-canonicalized path. I'm not actually quite sure why we don't use that function to discover the sysroot in librustc_session (cc @bjorn3 since you modified the sysroot code recently), but instead just call the current_exe-based function.

I'm noy

**Comment by bjorn3** [comment]
If I run `rustc a.rs`, `env::args().next()` would return `rustc`, right? In that case it will look for the sysroot in `../`, which is unlikely to be the real sysroot location.

> However, since it can be set to an arbitrary value, fs::get_or_default_sysroot() and its return value should not be relied on for security-related decisions.

Rustc runs with the same permissions as the user. I personally don't see how security is relevant here. It can only be invoked by the user. Also the sysroot is most of the time only used to load the files the crate will link against. Only with `-Zcodegen-backend` can it dynamically link against a dylib in the sysroot.

**Comment by bjorn3** [comment]
`sysroot_candidates` calls `get_or_default_sysroot`. The former also searches relative to the rustc_driver dylib for custom drivers, while the later only searches relative to the main executable. `sysroot_candidates` was necessary when cg_llvm was a hotpluggable codegen backend. Otherwise custom drivers wouldn't find the llvm codegen backend.

**Comment by rcvalle** [comment]
> If I run `rustc a.rs`, `env::args().next()` would return `rustc`, right? In that case it will look for the sysroot in `../`, which is unlikely to be the real sysroot location.
Indeed. We could perhaps check if sysroot is found using `env::args().next()`, and if it's not found, use `env::current_exe()`. What do you think?

> 
> > However, since it can be set to an arbitrary value, fs::get_or_default_sysroot() and its return value should not be relied on for security-related decisions.
> 
> Rustc runs with the same permissions as the user. I personally don't see how security is relevant here. It can only be invoked by the user. Also the sysroot is most of the time only used to load the files the crate will link against. Only with `-Zcodegen-backend` can it dynamically link against a dylib in the sysroot.
It may be relevant in a multiuser or remote system, where the rustc binary may be executed by a different user with different privileges.

**Comment by rcvalle** [comment]
@bjorn3 @Mark-Simulacrum I've amended my commit to check if sysroot is found using env::args().next(), and if it's not found, use env::current_exe() instead. What do you think?

File: compiler/rustc_session/src/filesearch.rs [hunk 1]
```diff
@@ -123,15 +123,44 @@ pub fn get_or_default_sysroot() -> PathBuf {
...
+            Err(e) => panic!("failed to get current_exe: {}", e),
+        }
+    }
+
+    // Use env::args().next() instead of env::current_exe() to get the path of
+    // the executable without following symlinks/canonicalizing any component.
+    // This makes the rustc binary able to locate Rust libraries in systems
+    // using content-addressable storage (CAS). However, since it can be set to
+    // an arbitrary value, fs::get_or_default_sysroot() and its return value
+    // should not be relied on for security-related decisions.
```
Comment (Mark-Simulacrum) [inline compiler/rustc_session/src/filesearch.rs:145]: This last sentence seems misleading, as it applies just as much to from_current_exe. I would perhaps expect to see it on the get_or_default_sysroot function itself.

**Comment by Mark-Simulacrum** [comment]
I am still concerned about this change. It seems like it could easily lead to confusion if there are (by accident or intentionally) different sysroots at argv0 and current_exe. That said, it does seem like the problem you're running into is a real one - I just am not sure this is the right solution. Can you say more about how other compilers or programs that need more than one file deal with this in your system? I think overall I just don't have enough expertise to weigh in here. cc @rust-lang/compiler - maybe someone can chime in if they have experience or thoughts here.

**Comment by wesleywiser** [comment]
Perhaps we should consider adding a flag that would allow specifying the sysroot path instead?

**Comment by jyn514** [comment]
@wesleywiser there's already a `--sysroot` flag:

```
rustc --help -v | grep sysroot
        --print [crate-name|file-names|sysroot|target-libdir|cfg|target-list|target-cpus|target-features|relocation-models|code-models|tls-models|target-spec-json|native-static-libs]
        --sysroot PATH  Override the system root
```

**Comment by wesleywiser** [comment]
Thanks @jyn514! GitHub hid [this comment](https://github.com/rust-lang/rust/pull/79253#issuecomment-732641000) which has the motivation for the change.

**Comment by jyn514** [comment]
@rcvalle would this be solved on your end by adding a wrapper script that passes the proper `--sysroot` to rustc?

**Comment by Mark-Simulacrum** [comment]
We discussed this in today's compiler team meeting, and raised the following:  

* we likely want some solution to this problem
* we are interested in whether --sysroot solves this (through wrapping or otherwise driving rustc)
* what do other compilers/programs with sysroots do?

I'm marking this as waiting on author until we get answers to these questions.

File: compiler/rustc_session/src/filesearch.rs (hunk 1 above)
Comment (rcvalle) [reply compiler/rustc_session/src/filesearch.rs]: Done. (I improved the changes a bit and fixed the comments.)

**Comment by rcvalle** [comment]
> We discussed this in today's compiler team meeting, and raised the following:
> 
>     * we likely want some solution to this problem
> 
>     * we are interested in whether --sysroot solves this (through wrapping or otherwise driving rustc)
@wesleywiser @jyn514 @Mark-Simulacrum Unfortunately not. `filesearch::get_or_default_sysroot()` is called directly in some locations regardless of what is specified using --sysroot, such as in https://github.com/rust-lang/rust/blob/5986dd878f3e432025eb1946149e3241d3998b1b/compiler/rustc_codegen_ssa/src/back/link.rs#L890 (and I don't know is it's an issue itself).
> 
>     * what do other compilers/programs with sysroots do?
The other compilers use include and library paths that can be specified instead of a sysroot.

> 
> 
> I'm marking this as waiting on author until we get answers to these questions.

**Comment by jyn514** [comment]
@rcvalle that should give precedencence to --sysroot; if not that's a bug and I would rather fix it directly than add hacks on top.

Assuming --sysroot works correctly, would that meet your use case? It sounds like for other compilers you're passing flags explicitly, which you can also do for rustc.

**Comment by rcvalle** [comment]
> @rcvalle that should give precedencence to --sysroot; if not that's a bug and I would rather fix it directly than add hacks on top.
I think these are separate:
1) What is the default compiler behavior (i.e., imply sysroot from the executable path following symlinks/canonicalizing components only, or try to find sysroot without following symlinks/canonicalizing components, and if is not found, use the former).
2) `--sysroot` option not being honored across rustc.

> 
> Assuming --sysroot works correctly, would that meet your use case? It sounds like for other compilers you're passing flags explicitly, which you can also do for rustc.
I can certainly look at 2 separately and test if it resolves the issue, but I think we should also define 1.

**Comment by rcvalle** [comment]
For reference, https://www.gnu.org/prep/standards/standards.html#Finding-Program-Files describes the behavior of conforming GNU programs (which should include GCC).

**Comment by Mark-Simulacrum** [comment]
I think matching the gcc/gnu docs in this aspect makes sense, though I would include env::current_exe after failing to find in argv0.

I think the easiest "is this a sysroot" identification is probably (given the supposed rustc binary path) a check that ../lib/rustlib is a directory. That seems like it should be true on your system and any standard Rust installation. Probably this is already somewhat defined in filesearch.rs but we should make sure.

That said, I believe fixing (2) to make sure we respect --sysroot is perhaps the better approach, since it'll give us results more reliably.

**Comment by rcvalle** [comment]
> I think matching the gcc/gnu docs in this aspect makes sense, though I would include env::current_exe after failing to find in argv0.
> 
> 
> 
> I think the easiest "is this a sysroot" identification is probably (given the supposed rustc binary path) a check that ../lib/rustlib is a directory. That seems like it should be true on your system and any standard Rust installation. Probably this is already somewhat defined in filesearch.rs but we should make sure.

This is exactly what these changes are doing. Please take another look when you have a chance and let me know what you think.

> 
> 
> 
> That said, I believe fixing (2) to make sure we respect --sysroot is perhaps the better approach, since it'll give us results more reliably.

Sure. I'll take a look and send a separate PR to fix this issue.

> 
> 
> 
>

File: compiler/rustc_session/src/filesearch.rs [hunk 2]
```diff
@@ -113,6 +113,10 @@ pub fn make_target_lib_path(sysroot: &Path, target_triple: &str) -> PathBuf {
     sysroot.join(&relative_target_lib_path(sysroot, target_triple))
 }
 
+// This function checks if sysroot is found using env::args().next(), and if it
+// is not found, uses env::current_exe() to imply sysroot. However, since
+// env::args().next() can be set to an arbitrary value, this function and its
+// return value should not be relied on for security-related decisions.
```
Comment (bjorn3) [inline compiler/rustc_session/src/filesearch.rs:119]:
```suggestion
// is not found, uses env::current_exe() to imply sysroot.
```

A dynamically linked codegen backend can be loaded from the sysroot, so we have to rely on it for security-related decisions. This shouldn't be a problem though, as a program that can pass an arbitrary value for the first argument can likely run arbitrary code anyway.

File: compiler/rustc_session/src/filesearch.rs [hunk 3]
```diff
@@ -123,15 +125,41 @@ pub fn get_or_default_sysroot() -> PathBuf {
...
+            Err(e) => panic!("failed to get current_exe: {}", e),
+        }
+    }
+
+    // Use env::args().next() to get the path of the executable without
+    // following symlinks/canonicalizing any component. This makes the rustc
+    // binary able to locate Rust libraries in systems using content-addressable
+    // storage (CAS).
+    fn from_env_args_next() -> Option<PathBuf> {
+        match env::args().next() {
```
Comment (nagisa) [inline compiler/rustc_session/src/filesearch.rs:147]: Should use `args_os` in order to avoid issues related to non-utf8 paths.

**Comment by nagisa** [comment]
> I am still concerned about this change. It seems like it could easily lead to confusion if there are (by accident or intentionally) different sysroots at argv0 and current_exe.

This problem seems very real and sounds super likely to break people's existing use-cases I think. This is especially because a candidate as PR is currently written is considered to be a candidate as soon as it contains a `lib{,32,64}` directory. Here's how it could go wrong. Consider these following directory structure:

```
/banana/
    bin/rustc -> /peach/bin/rustc
    lib/
        stuff here, but not rustlib/
/peach/
    bin/rustc
    lib/
        some.rlib
        rustlib/
            <snip>
```

If I'm reading this PR right, `/banana` would be considered the sysroot if rustc is invoked via `/banana/bin/rustc`, This is a (breaking) change in behaviour from today where `/peach` would get selected.

---

With that in mind, the only way I can see this kind of change working out is to consider sysroot to be not a single directory but a list of locations to look at when looking for necessary components (libraries, binaries, etc.). This doesn't need to be reflected in the CLI, but it would also let to specify the a choice between `argv0` and `current_exe` based sysroots as a default without regressing (almost) any current use-cases.

**Comment by jyn514** [comment]
> consider sysroot to be not a single directory but a list of locations to look at when looking for necessary components (libraries, binaries, etc.). 

That sounds like the same thing suggested in https://github.com/rust-lang/rust/pull/80933, actually. Seems like a good change to me.

**Comment by jyn514** [comment]
I wonder if that would help with miri and clippy's sysroot woes: https://github.com/rust-lang/rust/pull/78926#issuecomment-726290002

**Comment by rcvalle** [comment]
> If I'm reading this PR right, `/banana` would be considered the sysroot if rustc is invoked via `/banana/bin/rustc`, This is a (breaking) change in behaviour from today where `/peach` would get selected. Its breaking because the sysroot at `/banana` is not actually a valid rust sysroot.
> 

In this case, "/peach" would still be selected because it uses find_libdir() to check for the existence of "rustlib/" inside the libdir. Let me know if you think it isn't enough, and if there are any possible scenarios where it would change the current behavior.

**Comment by nagisa** [comment]
> In this case, "/peach" would still be selected because it uses find_libdir() to check for the existence of "rustlib/" inside the libdir. Let me know if you think it isn't enough, and if there are any possible scenarios where it would change the current behavior.

Right, I didn't look at `find_libdir` closely enough. There is still a problem however, but significantly more convoluted (and less likely to come up in the wild, I guess). Consider a case where both `banana` and `peach` sysroots has a distinct sets of targets installed and so both contain a `lib/rustlib` but the contents of these are different and what you can do with them also differs.

**Comment by rcvalle** [comment]
> Right, I didn't look at `find_libdir` closely enough. There is still a problem however, but significantly more convoluted (and less likely to come up in the wild, I guess). Consider a case where both `banana` and `peach` sysroots has a distinct sets of targets installed and so both contain a `lib/rustlib` but the contents of these are different and what you can do with them also differs. 

Can you think if any straightforward way we could check for/disambiguate sysroot in such case? If this is a less likely case, couldn't the `--sysroot` option be used to disambiguate it? Admittedly not knowing of a use case, I'd argue this would be a environment setup/installation problem, and not a rustc issue.

**Comment by nagisa** [comment]
> Can you think if any straightforward way we could check for/disambiguate sysroot in such case?

I feel that between #80933 and this the approach that would receive least resistance is an implementation that searches for necessary things in multiple directiories in some order. Possibly `--sysroot -> argv0 -> current_exe` or some such.

We could still attempt to adjust `find_libdir` to check for the target directory within `rustlib/` but that seems… meh? And you would still need to implement some sort of search/fallback mechanism for #80933 _anyway_.

**Comment by rcvalle** [comment]
> I feel that between #80933 and this the approach that would receive least resistance is an implementation that searches for necessary things in multiple directiories in some order. Possibly `--sysroot -> argv0 -> current_exe` or some such. That however seems like it is inching closer to the [MCP area](https://forge.rust-lang.org/compiler/mcp.html).
> 

It seems that the decision between looking into sysroot candidates or using the sysroot specified using the sysroot option is made before calling fs::get_or_default_sysroot(), and I don't intend to change that, since as you've pointed out it'd likely be a major change. That said, the rest of the implementation is what is in these changes.

> 
> 
> Could still attempt to adjust `find_libdir` to check for the target directory within `rustlib/` but that seems… meh? And somebody would still need to implement some sort of search/fallback mechanism for #80933 _anyway_.

For #80933, it's straightforward to check if the sanitizer libraries exist in the specified/session sysroot, and if not, use the default sysroot resulting from these changes (if merged).

Let me know what you think.

File: compiler/rustc_session/src/filesearch.rs (hunk 3 above)
Comment (rcvalle) [reply compiler/rustc_session/src/filesearch.rs]: Done.

**Comment by nagisa** [comment]
I'm uncomfortable approving this PR in its current form for the reasons outlined in previous comments. I don't believe we have a good way to quantify if anybody is relying on the current behaviour in weird ways and I don't want to be on the hook if peoples' use-cases break.

**Comment by rcvalle** [comment]
For #80933, I've amended my commit to implement the changes I mentioned earlier. What do you think?

**Comment by rcvalle** [comment]
To summarize what we've discussed so far:

* Without these changes:
  * Breaks a real scenario for large build systems using CAS with symlink forests.
  * Doesn't implement the same behavior described in any standard.
  * Doesn't make rustc consistent with any programs and possible any other compilers.
* With these changes:
  * Breaks a hypothetical convoluted scenario (or possibly others) for which we don't know a use case.
  * Implements the same behavior described in a well-known standard (i.e., https://www.gnu.org/prep/standards/standards.html#Finding-Program-Files).
  * Makes rustc consistent with conforming GNU programs (which should include GCC) and possibly other compilers.

Is there anything I can do to make these changes more "mergeable"?

**Comment by rcvalle** [comment]
@nagisa What if I change it to check if sysroot is found using env::args().next() **only if the rustc in argv[0] is a symlink**; otherwise or if it is not found, use env::current_exe() to imply sysroot? While I still prefer the behavior described in https://www.gnu.org/prep/standards/standards.html#Finding-Program-Files, it's a compromise that would fix the scenario for large build systems using CAS with symlink forests and wouldn't break the scenario you identified. What do you think?

**Comment by nagisa** [comment]
Hm, yeah I think that kind of approach could be compatible enough.

**Comment by rcvalle** [comment]
@nagisa I've amended my commit to implement the changes I mentioned earlier. Please take another look when you have a chance.

**Comment by nagisa** [comment]
Yeah, this seems reasonable enough. Thanks!

@bors r+

**Comment by rcvalle** [comment]
Thank you @nagisa, @Mark-Simulacrum, @jyn514, @bjorn3, and others for your time on this! Much appreciated.

**Pull request merged**